aiohttp/connector.py

```diff
@@ -470,25 +479,33 @@ def _get(self, key):
         del self._conns[key]
         return None
 
+    def _release_key_waiter(self, key, waiters):
+        if not waiters:
+            return False
+
+        waiter = waiters.pop(0)
+        waiter.set_result(None)
```

Comment:
You can not guarantee that the element is a none canceled future. Even you are removing from the list here [1].

Once the `Future` is set with an exception the callback [2] that will trigger this Exception will be scheduled by the loop, so before the callback that cancels the Task is really executed, the task that was holding a connection and wants to release it might `pop` a canceled future. 

[1] https://github.com/aio-libs/aiohttp/pull/2937/files#diff-7f25afde79f309e2f8722c26cf1f10adR399
[2] https://github.com/python/cpython/blob/master/Lib/asyncio/futures.py#L254

Reply:
I will mimic this code [1] https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L450

Reply:
I don't see how this can happen because the async stack ends on the waiter for `Future` and there are no done callbacks associated with it, see this example: https://ideone.com/izIfqw

Think of how this would work, currently no-one sets an exception on this future besides cancelling the waiter at [1], we can verify this by seeing the future is only available from `self._waiters` and no one sets an exception on it.

so, if you cancel task at [1], you have two options:

1) you don't wait on the cancelled task:
  1.1) you can release it (ok, removed from list), and then wait on cancelled (item already gone from list)
  1.2) you can not release it (ok), then wait on cancelled (item will be removed from list)
2) you wait on the cancelled task, in which case the baseexcept runs and removes item from list.  Nothing else can happen because we're just waiting on the future and no-one else is waiting on said future or has callbacks against said future.

Reply:
let me know if I'm missing something in my logic, I know it can get complex

Reply:
The task will end up once the reactor schedules again the task, the result that is given back to the task depends on the future taht yielded the task before. The cancel is only a signal that wakes up the task with an CanceledError exception.

So the task might be still there just waiting for its turn in the reactor/loop. While its happenning you have chances to have the situation that I told tou.

Please review the asyncio lock code, all of the different implementations take this into consideration.

Reply:
I think what's not clear perhaps, is that when you cancel the task the task is not marked cancelled until the task can process the event, in that case the waiter will be removed. The release cannot happen in this scenario because the removal code is at the bottom of the event chain, it could happen as the cancel bubbles up. Can you give a concrete example of your scenario, I still don't think it's possible.

Reply:
Better an image than thousand words

```python
import asyncio

waiters = []

async def task(waiter):
    try:
        await waiter
    finally:
        print("task finalitzed")
        if waiter in waiters:
            waiters.remove(waiter)
        else:
            print("waiter not present")

def wakeup_next_waiter():
    if not waiters:
        return

    waiter = waiters.pop(0)
    try:
        waiter.set_result(None)
    except Exception as e:
        print(f"Exception calling set_result {e!r}")
        raise

async def main(loop):
    waiter = loop.create_future()
    waiters.append(waiter)

    # create the task that will wait till either the waiter
    # is finised or the task is cancelled.
    t = asyncio.ensure_future(task(waiter))

    # make a loop iteration to allow the task reach the
    #     await waiter
    await asyncio.sleep(0)

    # put in the loop a callback to wake up the waiter.
    loop.call_soon(wakeup_next_waiter)

    # cancel the task, this will mark the task as cancelled
    # but will be pending a loop iteration to wake up the
    # task, having as a result a CanceledError exception.
    # This implicitly will schedule the Task._wakeup function
    # to be executed in the next loop iteration.
    t.cancel()

    try:
        await t
    except asyncio.CancelledError:
        pass

loop = asyncio.get_event_loop()
loop.run_until_complete(main(loop))
```

The output shows how `waiter.set_result(None)` statment raises an exception, the task never had the chance to execute the tear down code that must remove its waiter dependency.

So, the solution is the same applied in many other similiar situations in the asyncio lock implementation, check always [1] if the waiter has been finised or not.

[1] https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L450

Reply:
thanks @pfreixes for that example, so it seems it depends on the ordering of the tasks called during cancellation. Here's a question, can you have an outstanding release for a single connector?  It seems like you could end up releasing two connectors for a single cancel no if the except clause executes before `wakeup_next_waiter`

Reply:
Not sure about the question, but once the race condition is considered and mitigated the code should work as is expected.

Reply:
@pfreixes my "picture" :)

```python
import asyncio

loop = asyncio.get_event_loop()
waiters = []


async def task(waiter):
    try:
        await waiter
    except:
        print("task finalized")
        if waiter in waiters:
            print("waiter removed")
            try:
                waiters.remove(waiter)
            except ValueError:
                print("Waiter already removed")
        else:
            print("waiter not present")
        raise


def wakeup_next_waiter():
    if not waiters:
        return

    waiter = waiters.pop(0)
    try:
        if not waiter.done():
            waiter.set_result(None)
    except Exception as e:
        print(f"Exception calling set_result {e!r}")
        raise


async def main(loop):
    # We add two waiters
    waiter = loop.create_future()
    waiters.append(waiter)

    waiter = loop.create_future()
    waiters.append(waiter)

    # create the task that will wait till either the waiter
    # is finished or the task is cancelled.
    t = asyncio.ensure_future(task(waiter))

    # make a loop iteration to allow the task reach the
    #     await waiter
    await asyncio.sleep(0)

    # put in the loop a callback to wake up the waiter.
    loop.call_later(0.1, wakeup_next_waiter)

    # cancel the task, this will mark the task as cancelled
    # but will be pending a loop iteration to wake up the
    # task, having as a result a CanceledError exception.
    # This implicitly will schedule the Task._wakeup function
    # to be executed in the next loop iteration.
    t.cancel()

    try:
        await t
    except asyncio.CancelledError:
        pass

    # wait for the release to run
    await asyncio.sleep(1)

    # now we have zero waiters even though only one was cancelled
    print(len(waiters))

loop.run_until_complete(main(loop))
```

note in this example I add two waiters, cancel one, but at the end none are left because I ensure that the `wakeup_next_waiter`  happens after `task`

Reply:
I don't see the issue in the code - be carefull with extra ifs that are not needed in thte `task` function. If you are claiming that both waiters are removed this IMO is perfectly fine.

The first waiter that is created is removed through the happy path of the `wakeup_next_waiter`, obviously being removed by `waiter = waiters.pop(0)`, no task is wakened up because you didn't start any task with that waiter. The second waiter is canceled and the task related will remove the waiter by the `waiters.remove(waiter)`. 

Just to put all of us in the same page, the release of a used connection is done automatically by the context manager of a request calling the `resp.release()` [1]. So every time that a code path goes out of the context will release the connection and will try to wake up pending requests.

[1] https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client_reqrep.py#L786